src/backends/interflop-cancellation/interflop_cancellation.c

```diff
@@ -0,0 +1,280 @@
+ *  You should have received a copy of the GNU General Public License        *
+ *  along with Verificarlo.  If not, see <http://www.gnu.org/licenses/>.     *
+ *                                                                           *
+ *****************************************************************************/
+
+// Changelog:
+
+#include <argp.h>
+#include <err.h>
+#include <errno.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <sys/time.h>
+#include <unistd.h>
+#include <string.h>
+
+#include "../../common/float_const.h"
+#include "../../common/interflop.h"
+#include "../../common/tinymt64.h"
+
+#include "quadmath-imp.h"
+
+typedef struct {
+  int choose_seed;
+  uint64_t seed;
+} t_context;
+
+/* define default environment variables and default parameters */
+#define MCA_TOLERANCE_DEFAULT 1
+#define WARNING_DEFAULT 0
+
+static int WARN = WARNING_DEFAULT;
+
+static int MCALIB_T = MCA_TOLERANCE_DEFAULT;
+
+// possible op values
+#define MCA_ADD 1
+#define MCA_SUB 2
+#define MCA_MUL 3
+#define MCA_DIV 4
+
+#define min(a, b) ((a) < (b) ? (a) : (b))
+#define max(a, b) ((a) > (b) ? (a) : (b))
+#define bit(A,i) ((A >> i) & 1)
+#define bit_a_b(X,a,b) (((X << (sizeof(X)*8 - max(a,b) - 1)) >> (sizeof(X)*8 - max(a,b) - 1)) & ((X >> min(a,b) ) << min(a,b)))
+
+static float _mca_sbin(float a, float b, int qop);
+static double _mca_dbin(double a, double b, int qop);
+
+/******************** MCA CONTROL FUNCTIONS *******************
+ * The following functions are used to set virtual precision and
+ * MCA mode of operation.
+ ***************************************************************/
+
+static int _set_mca_tolerance(int tolerance) {
+  MCALIB_T = tolerance;
+  return 0;
+}
+
+static int _set_warning() {
+  WARN = 1;
+}
+
+/******************** MCA RANDOM FUNCTIONS ********************
+ * The following functions are used to calculate the random
+ * perturbations used for MCA
+ ***************************************************************/
+
+/* random generator internal state */
+static tinymt64_t random_state;
+
+static double _mca_rand(void) {
+  /* Returns a random double in the (0,1) open interval */
+  return tinymt64_generate_doubleOO(&random_state);
+}
+
+#define exp(X) ({                                     \
+  /* Convert X to an int */                           \
+  u_int64_t _V_ = *(u_int64_t*)&X;                    \
+  /* Get size of the exponent */                      \
+  int _E_ = (6+3*(sizeof(X) >> 2)-1);                 \ 
+  /* Get the size of the mantiss */                   \
+  int _M_ = (sizeof(X)*8-_E_-1);                      \
+  /* Compute the exponent */                          \     
+  (int)bit_a_b(_V_>>_M_,0,_E_-1) - ((1<<_E_-1)-2);})
+
+#define inexact(X,S) (X + pow(2,exp(X)-S) * _mca_rand())
+
+static void _set_mca_seed(int choose_seed, uint64_t seed) {
```

Comment:
This will have to be refactored after merging #152

(now there is a standard function in `src/common`)